### PR TITLE
Expand AEAD limits to consider multi-user security.

### PR DESCRIPTION
**Pull request by chris-wood**
In a typical multi-user setting, the adversary is assumed to perform some
(massive) amount of offline work to break the integrity of a single random
connection. Each user is represented as a unique (key, nonce) pair. In
considering the same threat model, we must treat each key resulting from
a KeyUpdate event as a unique user. As a result, any bounds implied from
the multi-user setting imply (minimally) for the lifetime of a single
connection.

Hoang et al. present tight multi-user security bounds for randomized AES-GCM
(as is used in TLS 1.3 and QUIC), so we can take advantage of those for
per-connection limits. In contrast, there are no multi-user security bounds
for AEAD_CHACHA20_POLY1305 or AEAD_AES_128_CCM, so we must use the
single-user bounds in their stead.

cc @martinthomson, @fxguenther

### CODE REVIEW

**Comment by chris-wood** [comment]
Note: this is just a draft to get us started!

File: draft-ietf-quic-tls.md [hunk 1]
```diff
@@ -100,6 +100,15 @@ informative:
...
     target: "https://www.felixguenther.info/docs/QUIPS2020_RobustChannels.pdf"
 
+  MultiUserSecurity:
+    title: "The Multi-user Security of GCM, Revisited: Tight Bounds for Nonce Randomization"
+    author:
+      - ins: V. T. Hoang
+      - ins: S. Tessaro
+      - ins: A. Thiruvengadam
+    date: 2018-10-18
+    target: https://dl.acm.org/doi/10.1145/3243734.3243816
```
Comment (martinthomson) [inline draft-ietf-quic-tls.md]: You can cite this using `{{?GCM-MU=DOI.10.1145/3243734.3243816}}` inline.

File: draft-ietf-quic-tls.md [hunk 2]
```diff
@@ -1552,30 +1561,39 @@ number of attempts to forge packets. TLS achieves this by closing connections
...
-  packet protection keys.  To avoid exhaustion of both sets of keys, it might be
-  necessary to initiate two key updates in succession.
-
-For AEAD_AES_128_GCM, AEAD_AES_256_GCM, and AEAD_CHACHA20_POLY1305, the limit on
-the number of packets that fail authentication is 2^36.  Note that the analysis
-in {{AEBounds}} supports a higher limit for the AEAD_AES_128_GCM and
-AEAD_AES_256_GCM, but this specification recommends a lower limit.  For
-AEAD_AES_128_CCM, the limit on the number of packets that fail authentication
-is 2^23.5; see {{ccm-bounds}}.
+: Due to the way that header protection protects the Key Phase, packets that
```
Comment (martinthomson) [inline draft-ietf-quic-tls.md]: This looks like a rewrapping only.  Maybe revert that.

File: draft-ietf-quic-tls.md [hunk 3]
```diff
@@ -1552,30 +1561,39 @@ number of attempts to forge packets. TLS achieves this by closing connections
...
-is able to successfully forge a packet; see {{AEBounds}} and {{ROBUST}}.
+Endpoints MUST count the number of received packets that fail authentication
+for each set of keys and during the lifetime of a connection. If the number of
+packets that fail authentication with the same key exceeds a limit that is
+specific to the AEAD in use, the endpoint MUST stop using those keys.
+Endpoints MUST initiate a key update before reaching this limit.
+If a key update is not possible, the endpoint MUST immediately close the
+connection. If the number of packets that fail authentication within the
+connection (across all keys) exceeds a limit that is specific to the AEAD in
+use, the endpoint MUST immediately close the connection. Applying these limits
```
Comment (martinthomson) [inline draft-ietf-quic-tls.md]: An immediate close can still involve sending and receiving packets, so you need to add " and not attempt to process any further packets".  Or something along these lines.

File: draft-ietf-quic-tls.md [hunk 4]
```diff
@@ -2170,10 +2188,10 @@ packet = 4cfe4189655e5cd55c41f69080575d7999c25a5bfb
...
 # Analysis of Limits on AEAD_AES_128_CCM Usage {#ccm-bounds}
 
-TLS {{?TLS13}} and {{AEBounds}} do not specify limits on usage for
-AEAD_AES_128_CCM. However, any AEAD that is used with QUIC requires limits on
-use that ensure that both confidentiality and integrity are preserved. This
-section documents that analysis.
+TLS {{?TLS13}} and {{AEBounds}} do not specify per-key or per-connection
+limits on usage for AEAD_AES_128_CCM. However, any AEAD that is used with QUIC
+requires limits on use that ensure that both confidentiality and integrity are
+preserved. This section documents that analysis.
```
Comment (martinthomson) [inline draft-ietf-quic-tls.md]: Perhaps change to: "This section documents an analysis of per-key limits, which - absent an analysis of multi-user security - also set per-connection limits."

**Review by martinthomson — COMMENTED** [review]
I have to admit that I don't really understand how this change is motivated.  How is use of disparate keys within the same connection any different than use of different keys in multiple connections?

**Comment by chris-wood** [comment]
> How is use of disparate keys within the same connection any different than use of different keys in multiple connections?

It's not, really. As I understand it, we're accepting we don't have a global view of all users, and thus limit our view of "users" to "all keys in a given connection." As a result, it's possible that these limits are higher than we might expect in a typical (global) multi-user setting, but I'm not sure we can do much about that beyond bumping the block (and key) size.

File: draft-ietf-quic-tls.md [hunk 5]
```diff
@@ -1570,12 +1564,14 @@ Note:
...
-AEAD_AES_128_CCM, the limit on the number of packets that fail authentication
-is 2^23.5; see {{ccm-bounds}}.
+Key updates MUST be initiated before usage limits on packet protection keys are
+exceeded. For AEAD_AES_128_GCM and AEAD_AES_256_GCM, the confidentiality limit
+is 2^27 encrypted packets, and the integrity limit is 2^54 encrypted or forged
+packets; see {{?GCM-MU}} and {{gcm-bounds}}. For AEAD_CHACHA20_POLY1305, the
+limit on number of packets that fail authentication is 2^93; see {{AEBounds}}.
+For AEAD_AES_128_CCM, the confidentiality limit is 2^23 encrypted packets, and
+the integrity limit is 2^23.5 forged packets; see {{ccm-bounds}}. Future
+analyses and specifications MAY relax these limits.
```
Comment (martinthomson) [inline draft-ietf-quic-tls.md]: We probably need to split this paragraph now.  One for the text on key updates: update before you get to L[confidentiality].  And a separate one for stop using the connection when you get to L[integrity].

File: draft-ietf-quic-tls.md [hunk 6]
```diff
@@ -2220,6 +2208,66 @@ is, `2l = 2^11`). This simplification is based on the packet containing all of
...
+an advantage in successfully forging a packet of no more than:
+
+~~~
+(1 / 2^(8*n)) + ((2 * v) / 2^2*n) + ((2 * o * v) / 2^(k + n))
+        + (n * (v + (v * l)) / 2^k)
+~~~
+
+The goal is to limit this advantage to 2^-57, to match the target in
+{{?TLS13}}. In this inequality, the last term dominates the rest, so the others
+can be removed without affecting the result. This produces the approximation:
```
Comment (martinthomson) [inline draft-ietf-quic-tls.md]: We should probably add a new limit for AEAD_AES_256_GCM.  That is, keep `k` in the following equation and then say 2^54 for 128-bit keys, and then "something much larger" for 256-bit keys, but setting a limit of 2^54 is probably good for that too.  (Honestly, if you have discarded that many packets and are still going, good for you.)

**Review by martinthomson — COMMENTED** [review]
This is pretty good.  I know that the analysis in the appendix is going to cause some concerns, but it's better than what we previously had.

File: draft-ietf-quic-tls.md [hunk 7]
```diff
@@ -2209,10 +2195,10 @@ v:
...
   updating keys.
 
-The analysis of AEAD_AES_128_CCM relies on a count of the number of block
-operations involved in producing each message. For simplicity, and to match the
-analysis of other AEAD functions in {{AEBounds}}, this analysis assumes a
-packet length of 2^10 blocks and a packet size limit of 2^14.
+The analyses that follow rely on a count of the number of block operations involved
+in producing each message. For simplicity, and to match the analysis of other AEAD
+functions in {{AEBounds}}, this analysis assumes a packet length of 2^10 blocks
+and a packet size limit of 2^14.
```
Comment (martinthomson) [inline draft-ietf-quic-tls.md]:
```suggestion
functions in {{AEBounds}}, this analysis assumes a packet length of 2^10 blocks;
that is, a packet size limit of 2^14 bytes.
```

File: draft-ietf-quic-tls.md [hunk 8]
```diff
@@ -2222,6 +2208,66 @@ is, `2l = 2^11`). This simplification is based on the packet containing all of
...
+{{?GCM-MU}} specify concrete bounds for AEAD_AES_128_GCM and AEAD_AES_256_GCM as
+used in TLS 1.3 and QUIC. This section documents this analysis using several
+simplifying assumptions:
+
+- The number of ciphertext blocks an attacker uses in forgery attempts is bounded
+by v * l, the number of forgery attempts and the size of each packet (in blocks).
+
+- The amount of offline work done by an attacker does not dominate other factors
+in the analysis.
+
```
Comment (martinthomson) [inline draft-ietf-quic-tls.md:2236]:
```suggestion

The bounds in {{?GCM-MU}} are tighter and more complete than those used in
{{AEBounds}}, which allows for larger limits than those described in {{?TLS13}}.
```

File: draft-ietf-quic-tls.md (hunk 5 above)
Comment (martinthomson) [reply draft-ietf-quic-tls.md]: How about:

```suggestion
Key updates MUST be initiated before confidentiality limits for the selected
AEAD are exceeded.  For AEAD_AES_128_GCM and AEAD_AES_256_GCM, the
confidentiality limit is 2^27 encrypted packets; see {{gcm-bounds}}.  For
AEAD_CHACHA20_POLY1305, the confidentiality limit is greater than the number of
possible packets (2^62) and so can be disregarded.  For AEAD_AES_128_CCM, the
confidentiality limit is 2^23 encrypted packets; see {{ccm-bounds}}.

A connection MUST be closed if the number of packets that fail authentication
exceed the integrity limit for the selected AEAD. For AEAD_AES_128_GCM, and the
integrity limit is 2^54 encrypted or forged packets; see {{gcm-bounds}}.  For
AEAD_AES_256_GCM, the integrity limit For AEAD_CHACHA20_POLY1305, the limit on
number of packets that fail authentication is 2^36; see {{AEBounds}}.  For
AEAD_AES_128_CCM, the integrity limit is 2^23.5 forged packets; see
{{ccm-bounds}}.

Future analyses and specifications MAY relax confidentiality or integrity limits
for an AEAD.
```

I changed the 2^93 to 2^36 for ChaCha.  I think that the 93 was in error.

File: draft-ietf-quic-tls.md [hunk 9]
```diff
@@ -2229,6 +2229,8 @@ by v * l, the number of forgery attempts and the size of each packet (in blocks)
 - The amount of offline work done by an attacker does not dominate other factors
 in the analysis.
 
+The bounds in {{?GCM-MU}} are tighter and more complete than those used in
+{{AEBounds}}, which allows for larger limits than those described in {{?TLS13}}.
 ### Confidentiality Limit
```
Comment (martinthomson) [inline draft-ietf-quic-tls.md:2240]:
```suggestion

### Confidentiality Limit
```
Just in case.

File: draft-ietf-quic-tls.md (hunk 6 above)
Comment (chris-wood) [reply draft-ietf-quic-tls.md]: Yeah, let's use the same limit for both, but show (in the appendix) the derivation of the k=256 case. Stand by for an update.

Comment (chris-wood) [reply draft-ietf-quic-tls.md]: Okay, done!

File: draft-ietf-quic-tls.md (hunk 9 above)
Comment (chris-wood) [reply draft-ietf-quic-tls.md:2240]: Already done.

**Review by martinthomson — APPROVED** [review]
This should cause some consternation when we introduce it, but I don't think that people will get too badly steamed.  The math is sound, at least to the extent that the underlying analyses are sound :)  And it's pretty directly actionable.

**Comment by chris-wood** [comment]
Thanks, @martinthomson! I'll draft an issue for the main repo and share it soon.

File: draft-ietf-quic-tls.md [hunk 10]
```diff
@@ -2220,6 +2221,81 @@ is, `2l = 2^11`). This simplification is based on the packet containing all of
...
+
+For AEAD_AES_256_GCM, the second and fourth terms dominate the rest, so the
+others can be removed without affecting the result. This produces the following
+approximation:
+
+~~~
+v <= 2^182
+~~~
+
+This is substantially smaller than the limit for AEAD_AES_128_GCM. Thus,
```
Comment (martinthomson) [inline draft-ietf-quic-tls.md]:
```suggestion
This is substantially smaller than the limit for AEAD_AES_128_GCM. However,
```

File: draft-ietf-quic-tls.md [hunk 11]
```diff
@@ -2220,6 +2221,81 @@ is, `2l = 2^11`). This simplification is based on the packet containing all of
...
+for consistency, this document uses the larger of the two -- 2^54 -- for both.
```
Comment (martinthomson) [inline draft-ietf-quic-tls.md]:
```suggestion
this document recommends that the smaller limit from AEAD_AES_128_GCM also
apply to AEAD_AES_256_GCM.
```

(The only important bit being s/larger/smaller/.)

This is partly consistency (because it is easier to use a single number here) and partly being conservative (because that large number is just unbelievable), and partly because that smaller value is already large enough.

**Review by martinthomson — APPROVED** [review]
I keep finding things, sorry.

File: draft-ietf-quic-tls.md (hunk 11 above)
Comment (chris-wood) [reply draft-ietf-quic-tls.md]: Woof. That one was bad. Good catch!

File: draft-ietf-quic-tls.md [hunk 12]
```diff
@@ -1570,12 +1564,23 @@ Note:
...
+AEAD are exceeded.  For AEAD_AES_128_GCM and AEAD_AES_256_GCM, the
+confidentiality limit is 2^27 encrypted packets; see {{gcm-bounds}}.  For
+AEAD_CHACHA20_POLY1305, the confidentiality limit is greater than the number of
+possible packets (2^62) and so can be disregarded.  For AEAD_AES_128_CCM, the
+confidentiality limit is 2^23 encrypted packets; see {{ccm-bounds}}.
+
+A connection MUST be closed if the number of packets that fail authentication
+exceed the integrity limit for the selected AEAD. For AEAD_AES_128_GCM, and the
+integrity limit is 2^54 encrypted or forged packets; see {{gcm-bounds}}.  For
+AEAD_AES_256_GCM, the integrity limit For AEAD_CHACHA20_POLY1305, the limit on
```
Comment (ekr) [inline draft-ietf-quic-tls.md]: Something is wrong with this text.

File: draft-ietf-quic-tls.md [hunk 13]
```diff
@@ -2220,6 +2221,82 @@ is, `2l = 2^11`). This simplification is based on the packet containing all of
...
+
+For AEAD_AES_256_GCM, the second and fourth terms dominate the rest, so the
+others can be removed without affecting the result. This produces the following
+approximation:
+
+~~~
+v <= 2^182
+~~~
+
+This is substantially smaller than the limit for AEAD_AES_128_GCM. However,
```
Comment (ekr) [inline draft-ietf-quic-tls.md]: Larger?

Comment (chris-wood) [reply draft-ietf-quic-tls.md]: Fixed.

File: draft-ietf-quic-tls.md (hunk 12 above)
Comment (chris-wood) [reply draft-ietf-quic-tls.md]: Fixed -- thanks!

File: draft-ietf-quic-tls.md [hunk 14]
```diff
@@ -1570,12 +1564,23 @@ Note:
   packet protection keys.  To avoid exhaustion of both sets of keys, it might be
   necessary to initiate two key updates in succession.
 
-For AEAD_AES_128_GCM, AEAD_AES_256_GCM, and AEAD_CHACHA20_POLY1305, the limit on
-the number of packets that fail authentication is 2^36.  Note that the analysis
-in {{AEBounds}} supports a higher limit for the AEAD_AES_128_GCM and
-AEAD_AES_256_GCM, but this specification recommends a lower limit.  For
-AEAD_AES_128_CCM, the limit on the number of packets that fail authentication
-is 2^23.5; see {{ccm-bounds}}.
+A connection MUST be closed if the number of encrypted packets exceeds the
```
Comment (martinthomson) [inline draft-ietf-quic-tls.md]: I missed this last time around.

I think that this is unrealistic.  People will make new connections, which have the same properties as new keys.  So we should just recommend key updates.

**Review by martinthomson — CHANGES_REQUESTED** [review]

File: draft-ietf-quic-tls.md (hunk 14 above)
Comment (chris-wood) [reply draft-ietf-quic-tls.md]: Hmm, fair point. They are functionally the same thing since we're ignoring the number of users. That said, doesn't it seem inconsistent to drop connections in the case of integrity, but then re-key for confidentiality?

File: draft-ietf-quic-tls.md [hunk 15]
```diff
@@ -1538,27 +1538,21 @@ discarded.
...
+In addition to counting packets sent, endpoints MUST count the number of
+received packets that fail authentication during the lifetime of a connection.
+If the number of packets sent or fail authentication within the connection
+(across all keys) exceeds a limit that is specific to the AEAD in use, the
+endpoint MUST immediately close the connection and not attempt to process any
+further packets. Applying a limit reduces the probability that an attacker can
+distinguish the AEAD in use from a random permutation; see {{AEBounds}},
+{{ROBUST}}, and {{?GCM-MU=DOI.10.1145/3243734.3243816}}.
 
 Note:
```
Comment (martinthomson) [inline draft-ietf-quic-tls.md]: I think that we can drop this note now that the failed authentications are counted globally.

File: draft-ietf-quic-tls.md [hunk 16]
```diff
@@ -1586,13 +1591,13 @@ Note:
...
-authentication.  Providing a reference to any analysis upon which values are
-based - and any assumptions used in that analysis - allows limits to be adapted
-to varying usage conditions.
+Any TLS cipher suite that is specified for use with QUIC MUST define
+per-connection limits on the use of the associated AEAD function that
+preserves margins for confidentiality and integrity. That is, limits MUST be
+specified for the number of packets that can be authenticated and for the
+number packets that can fail authentication. Providing a reference to any
+analysis upon which values are based - and any assumptions used in that
+analysis - allows limits to be adapted to varying usage conditions.
```
Comment (martinthomson) [inline draft-ietf-quic-tls.md]: And maybe revert this change.

File: draft-ietf-quic-tls.md [hunk 17]
```diff
@@ -1538,27 +1538,34 @@ discarded.
 
 ## Minimum Key Update Frequency
 
-Key updates MUST be initiated before usage limits on packet protection keys are
-exceeded. For the cipher suites mentioned in this document, the limits in
-Section 5.5 of {{!TLS13}} apply. {{!TLS13}} does not specify a limit for
-AEAD_AES_128_CCM, but the analysis in {{ccm-bounds}} shows that a limit of 2^23
-packets can be used to obtain the same confidentiality protection as the limits
-specified in TLS.
```
Comment (martinthomson) [inline draft-ietf-quic-tls.md:1546]: I didn't get enough time to do this today, but what I think we need here is some of the framing from this.  And that likely comes with a new section title and maybe even some restructuring.

File: draft-ietf-quic-tls.md (hunk 15 above)
Comment (chris-wood) [reply draft-ietf-quic-tls.md]: Good point -- dropped.

File: draft-ietf-quic-tls.md (hunk 16 above)
Comment (chris-wood) [reply draft-ietf-quic-tls.md]: Yep, done.

File: draft-ietf-quic-tls.md (hunk 17 above)
Comment (chris-wood) [reply draft-ietf-quic-tls.md:1546]: I reorganized things a bit. There's a parallel flow for the connection and integrity limit text now that think works. @ekr, what do you think of the new text?

File: draft-ietf-quic-tls.md [hunk 18]
```diff
@@ -1536,14 +1536,7 @@ After this period, old read keys and their corresponding secrets SHOULD be
...
-
-Key updates MUST be initiated before usage limits on packet protection keys are
-exceeded. For the cipher suites mentioned in this document, the limits in
-Section 5.5 of {{!TLS13}} apply. {{!TLS13}} does not specify a limit for
-AEAD_AES_128_CCM, but the analysis in {{ccm-bounds}} shows that a limit of 2^23
-packets can be used to obtain the same confidentiality protection as the limits
-specified in TLS.
+## Limits on Key Usage
 
 The usage limits defined in TLS 1.3 exist for protection against attacks
```
Comment (martinthomson) [inline draft-ietf-quic-tls.md:1545]: This still jumps in a bit cold.

```suggestion
This document sets usage limits for AEAD keys to ensure that overuse of a single key does
not give an adversary a disproportionate advantage in attacking the confidentiality
and integrity of communications.

The usage limits defined in TLS 1.3 exist for protection against attacks
```

Or some such nonsense.

**Review by martinthomson — COMMENTED** [review]
I'm much happier with this.  Thanks for fixing this up Chris.

File: draft-ietf-quic-tls.md (hunk 18 above)
Comment (chris-wood) [reply draft-ietf-quic-tls.md:1545]: I'm happy with that as an intro, though might tweak it a bit to emphasize that limits don't always apply to keys (but to connections).

File: draft-ietf-quic-tls.md [hunk 19]
```diff
@@ -1538,9 +1538,9 @@ discarded.
 
 ## Limits on Key Usage
```
Comment (martinthomson) [inline draft-ietf-quic-tls.md]: Like the tweak.  What do you think about the heading?

```suggestion
## Limits on AEAD Usage
```

**Review by martinthomson — APPROVED** [review]

File: draft-ietf-quic-tls.md (hunk 19 above)
Comment (chris-wood) [reply draft-ietf-quic-tls.md]: I originally left it as "Limits on Key Usage" for consistency with RFC8446, though I guess that's not the best rationale. I'm fine with the suggestion.
